transactor/src/main/scala/io/mediachain/copycat/TransactorService.scala

```diff
@@ -0,0 +1,132 @@
+package io.mediachain.copycat
+
+import java.util.concurrent.ExecutorService
+
+import cats.data.Xor
+import io.grpc.{ServerBuilder, Status, StatusRuntimeException}
+import io.mediachain.multihash.MultiHash
+import io.mediachain.protocol.CborSerialization
+import io.mediachain.protocol.Datastore._
+import io.mediachain.protocol.Transactor.JournalError
+import io.mediachain.protocol.transactor.Transactor.{InsertRequest, TransactorServiceGrpc, UpdateRequest}
+import io.mediachain.protocol.transactor.Transactor
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration.{Duration, SECONDS}
+
+class TransactorService(client: Client,
+                        timeout: Duration = Duration(120, SECONDS))
+                       (implicit val executionContext: ExecutionContext)
+  extends TransactorServiceGrpc.TransactorService {
+
+  override def fetchObjectChainHead(request: Transactor.MultihashReference):
+  Future[Transactor.MultihashReference] = {
+    val ref = MultiHash.fromBase58(request.reference)
+      .map(MultihashReference.apply)
+      .getOrElse {
+        throw new StatusRuntimeException(Status.INVALID_ARGUMENT)
+      }
+
+    client
+      .lookup(ref)
+      .map { (x: Option[Reference]) =>
+        x.map { ref =>
+          refToRPCMultihashRef(ref)
+        }.getOrElse {
+          throw new StatusRuntimeException(Status.NOT_FOUND)
+        }
+      }
+  }
+
+  private def refToRPCMultihashRef(ref: Reference)
+  : Transactor.MultihashReference = ref match {
+    case MultihashReference(multihash) =>
+      Transactor.MultihashReference(multihash.base58)
+    case r =>
+      throw new ClassCastException(
+        s"Expected MultihashReference, got type ${r.getClass.getTypeName}"
+      )
+  }
+
+  override def insertCanonicalRecord(request: InsertRequest)
+  : Future[Transactor.MultihashReference] = {
+    val recXor = CborSerialization.fromCborBytes[CanonicalRecord](
+      request.canonicalCbor.toByteArray
+    )
+
+    val insertF = recXor match {
+      case Xor.Left(err) => throw new StatusRuntimeException(
+        Status.INVALID_ARGUMENT.withDescription(
+          s"Object deserialization error: ${err.message}"
+        )
+      )
+      case Xor.Right(obj) =>
+        client.insert(obj)
```

Comment:
I think overall sanity checking is a separate discussion but object size check is probably good because we're writing to datastore directly from client